cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1520,28 +1520,42 @@ private Map<ProgramRunId, RunRecordDetail> getUnfinishedRuns(Set<ProgramRunId> p
     List<List<Field<?>>> allKeys = new ArrayList<>();
     for (ProgramRunId programRunId : programRunIds) {
       allKeys.add(getProgramRunInvertedTimeKey(TYPE_RUN_RECORD_ACTIVE, programRunId,
-                                               RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS)));
+                                               RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS),
+                                               false));
     }
-    return getRunRecordsTable().multiRead(allKeys).stream()
-      .map(AppMetadataStore::deserializeRunRecordMeta)
-      .collect(Collectors.toMap(RunRecordDetail::getProgramRunId, r -> r, (r1, r2) -> {
-        throw new IllegalStateException("Duplicate run record for " + r1.getProgramRunId());
-      }, LinkedHashMap::new));
+    
+    return getRangeScanRunsMap(allKeys);
   }
 
   private Map<ProgramRunId, RunRecordDetail> getCompletedRuns(Set<ProgramRunId> programRunIds) throws IOException {
     List<List<Field<?>>> allKeys = new ArrayList<>();
     for (ProgramRunId programRunId : programRunIds) {
-      List<Field<?>> keys = getRunRecordProgramPrefix(TYPE_RUN_RECORD_COMPLETED, programRunId.getParent());
+      // Get all keys without version
+      List<Field<?>> keysWithoutVersion = getRunRecordProgramPrefix(TYPE_RUN_RECORD_COMPLETED,
+                                                                    programRunId.getParent(),
+                                                                    false);
       // Get start time from RunId
       long programStartSecs = RunIds.getTime(RunIds.fromString(programRunId.getRun()), TimeUnit.SECONDS);
-      keys.add(Fields.longField(StoreDefinition.AppMetadataStore.RUN_START_TIME,
-                                getInvertedTsKeyPart(programStartSecs)));
-      keys.add(Fields.stringField(StoreDefinition.AppMetadataStore.RUN_FIELD, programRunId.getRun()));
-      allKeys.add(keys);
+      keysWithoutVersion.add(Fields.longField(StoreDefinition.AppMetadataStore.RUN_START_TIME,
+                                              getInvertedTsKeyPart(programStartSecs)));
+      keysWithoutVersion.add(Fields.stringField(StoreDefinition.AppMetadataStore.RUN_FIELD, programRunId.getRun()));
+      allKeys.add(keysWithoutVersion);
     }
-    return getRunRecordsTable().multiRead(allKeys).stream()
-      .map(AppMetadataStore::deserializeRunRecordMeta)
+
+    return getRangeScanRunsMap(allKeys);
+  }
+
+  private Map<ProgramRunId, RunRecordDetail> getRangeScanRunsMap(List<List<Field<?>>> allKeys) throws IOException {
```

Comment:
This method is a bit confusing. So this is for getting run record, just call it `getRunsByKeys`. Generally method name doesn't need to carry type (e.g. `Map` in this case).

Reply:
updated